### ISSUE TITLE
.match issues with TypeScript and fromCapture

### ISSUE THREAD

**Issue by j4k0xb** [codemod-js/codemod#898]
The `m.match` function doesn't seem to properly work with TypeScript. Example from the jsdoc:
```ts
import * as m from '@codemod/matchers';

let id: m.CapturedMatcher<t.Identifier>;
const idPlusIdMatcher = m.binaryExpression(
  '+',
  (id = m.capture(m.identifier())),
  m.fromCapture(id)
);

export default function() {
  return {
    BinaryExpression(path: NodePath<t.BinaryExpression>): void {
      m.match(idPlusIdMatcher, { id }, path.node, ({ id }) => {
        path.replaceWith(t.binaryExpression('*', id, t.numericLiteral(2)));
      });
    }
  };
}
```
![image](https://user-images.githubusercontent.com/55899582/216781267-e3496c83-5b1c-4945-a029-e8dec7ac3214.png)

This can be fixed by importing from `../matchers` instead, because the published package doesn't contain `src`.

https://github.com/codemod-js/codemod/blob/505da377a941d9a2ce3d6171e1a1bccff4042c4a/packages/matchers/src/utils/match.ts#L1
https://github.com/codemod-js/codemod/blob/505da377a941d9a2ce3d6171e1a1bccff4042c4a/packages/matchers/src/utils/matchPath.ts#L1

Another thing I noticed is that .fromCapture only works with captured primitives like strings, not matchers themselves:
```ts
// fails
const id = m.capture(m.identifier())
const idPlusIdMatcher = m.binaryExpression('+', id, m.fromCapture(id))
```
```ts
// works
const idValue = m.capture(m.anyString())
const idPlusIdMatcher = m.binaryExpression(
  '+',
  m.identifier(idValue),
  m.identifier(m.fromCapture(idValue))
)
```
Would be nice if both ways work to compare whole code structures

And when matching with containers it shouldn't narrow the type:
```ts
if (idPlusIdMatcher.match(ast)) {
  console.log(path.node) // TS says "t.BinaryExpression", at runtime it's e.g. "t.Program"
```

**Comment by eventualbuddha** [comment]
Fixed in v1.3.1.

**Comment by eventualbuddha** [comment]
And I also have an idea for the type narrowing thing. Stay tuned.
